pkg/core/blockchain.go

```diff
@@ -628,6 +629,122 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
+	}
+	sr, err := bc.stateRoot.GetStateRoot(height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve stateroot for height %d: %w", height, err)
+	}
+	v := bc.dao.Version
+
+	// Remove blocks/transactions/aers from currHeight down to height (not including height itself).
+	hHeight := bc.HeaderHeight()
+	cache := bc.dao.GetPrivate()
+	for i := height + 1; i <= currHeight; i++ {
+		err := cache.DeleteBlock(bc.headerHashes[i])
+		if err != nil {
+			return fmt.Errorf("error while removing block %d: %w", i, err)
+		}
+	}
+	for i := height + 1; i <= hHeight; i++ {
+		cache.DeleteHeader(bc.headerHashes[i])
+	}
+
+	// Completely remove contract IDs to update them later.
+	cache.Store.Seek(storage.SeekRange{Prefix: []byte{byte(storage.STContractID)}}, func(k, _ []byte) bool {
+		cache.Store.Delete(k)
+		return true
+	})
+
+	// Reset contracts storage and store new contract IDs.
+	var mode = mpt.ModeAll
+	if bc.config.RemoveUntraceableBlocks {
+		mode |= mpt.ModeGCFlag
+	}
+	// TODO: persist cache periodically, otherwise there's a risc of OOM.
+	s := mpt.NewTrieStore(sr.Root, mode, storage.NewPrivateMemCachedStore(cache.Store))
+	oldStoragePrefix := v.StoragePrefix
+	newStoragePrefix := statesync.TemporaryPrefix(oldStoragePrefix)
+	mgmtCSPrefixLen := 1 + 4 + 1 // STStorage + Management ID + contract state prefix
+	mgmtContractPrefix := make([]byte, mgmtCSPrefixLen-1)
+	id := int32(native.ManagementContractID)
+	binary.BigEndian.PutUint32(mgmtContractPrefix, uint32(id))
+	mgmtContractPrefix[4] = native.PrefixContract
+	cs := new(state.Contract)
+	s.Seek(storage.SeekRange{Prefix: []byte{byte(oldStoragePrefix)}}, func(k, v []byte) bool {
+		// May safely omit KV copying.
+		k[0] = byte(newStoragePrefix)
+		cache.Store.Put(k, v)
+
+		// TODO: remove this part after #2702.
+		if bytes.HasPrefix(k[1:], mgmtContractPrefix) {
+			var hash util.Uint160
+			copy(hash[:], k[mgmtCSPrefixLen:])
+			err = stackitem.DeserializeConvertible(v, cs)
+			if err != nil {
+				// TODO
+				panic(err)
+			}
+			// TODO: add convertabe compat test.
+			cache.PutContractID(cs.ID, hash)
+		}
+		return true
+	})
+	s.Close()
+
+	// Reset SYS-prefixed and IX-prefixed information.
+	cache.DeleteHeaderHashes(height+1, headerBatchCount)
+	cache.StoreAsCurrentBlock(b)
+	cache.PutCurrentHeader(b.Hash(), height)
+	v.StoragePrefix = newStoragePrefix
+	cache.PutVersion(v)
+
+	// Reset MPT.
+	err = bc.stateRoot.ResetState(height, cache.Store)
+	if err != nil {
+		return fmt.Errorf("failed to rollback MPT state: %w", err)
+	}
+
+	_, err = cache.Persist()
+	if err != nil {
+		return fmt.Errorf("failed to persist changes from cache to DAO: %w", err)
+	}
+	_, err = bc.dao.Persist()
+	if err != nil {
+		return fmt.Errorf("failed to persist changes: %w", err)
+	}
+
+	// Direct (cache-less) DB operations: reset transfer info and remove storage items.
+	// TODO: maybe we need to perform all operations using cache to avoid DB corrupts, if so, then need to refactor these two methods.
```

Comment:
Use state jump flags? It'd be nice to have this all work atomically, but I doubt it's practical enough, there can be a huge set of changes. Maybe we can reuse state jumping infrastructure a bit more.

Reply:
Yes, it can be reused, fixed.